modules/videoio/src/cap_images.cpp

```diff
@@ -184,77 +178,97 @@ bool CvCapture_Images::setProperty(int id, double value)
+    CV_Assert(offset);
 
-    if( !filename )
-        return 0;
+    *offset = 0;
 
     // check whether this is a valid image sequence filename
-    char *at = strchr(name, '%');
-    if(at)
+    std::string::size_type pos = filename.find('%');
+    if (pos != std::string::npos)
     {
-        unsigned int dummy;
-        if(sscanf(at + 1, "%ud", &dummy) != 1)
-            return 0;
-        name = strdup(filename);
+        pos++; CV_Assert(pos < len);
+        if (filename[pos] == '0') // optional zero prefix
+        {
+            pos++; CV_Assert(pos < len);
+        }
+        if (filename[pos] >= '1' && filename[pos] <= '9') // optional numeric size (1..9) (one symbol only)
+        {
+            pos++; CV_Assert(pos < len);
+        }
+        if (filename[pos] == 'd' || filename[pos] == 'u')
+        {
+            pos++;
+            if (pos == len)
+                return filename;  // end of string '...%5d'
+            CV_Assert(pos < len);
+            if (filename.find('%', pos) == std::string::npos)
+                return filename;  // no more patterns
+            CV_Error_(Error::StsBadArg, ("CAP_IMAGES: invalid multiple patterns: %s", filename.c_str()));
+        }
+        CV_Error_(Error::StsBadArg, ("CAP_IMAGES: error, expected '0?[1-9][du]' pattern, got: %s", filename.c_str()));
     }
     else // no pattern filename was given - extract the pattern
     {
-        at = name;
-
-        // ignore directory names
-        char *slash = strrchr(at, '/');
-        if (slash) at = slash + 1;
-
+        pos = filename.rfind('/');
 #ifdef _WIN32
-        slash = strrchr(at, '\\');
-        if (slash) at = slash + 1;
+        if (pos == std::string::npos)
+            pos = filename.rfind('\\');
 #endif
+        if (pos != std::string::npos)
+            pos++;
+        else
+            pos = 0;
 
-        while (*at && !isdigit(*at)) at++;
-
-        if(!*at)
-            return 0;
-
-        sscanf(at, "%u", offset);
+        while (pos < len && !isdigit(filename[pos])) pos++;
 
-        int size = (int)strlen(filename) + 20;
-        name = (char *)malloc(size);
-        CV_Assert(name != NULL);
-        strncpy(name, filename, at - filename);
-        name[at - filename] = 0;
-
-        strcat(name, "%0");
+        if (pos == len)
+        {
+            CV_Error_(Error::StsBadArg, ("CAP_IMAGES: can't find starting number (in the name of file): %s", filename.c_str()));
+        }
 
-        int i;
-        char *extension;
-        for(i = 0, extension = at; isdigit(at[i]); i++, extension++)
-            ;
-        char places[13] = {0};
-        sprintf(places, "%dd", i);
+        std::string::size_type pos0 = pos;
 
-        strcat(name, places);
-        strcat(name, extension);
+        std::string number_str;
+        while (pos < len && isdigit(filename[pos]))
+        {
+            number_str += filename[pos];
+            pos++;
+        }
+        CV_Assert(!number_str.empty());
```

Comment:
`... && number_str.size() <= 9);`? To fit 2,147,483,647 range, otherwise `atoi` can cause UB.

Reply:
`number_str` is used primary to capture 000000... prefix.

Reply:
removed `atoi()`, limit offset to 1e9